text/0000-local-default-generic-bounds.md

```diff
@@ -0,0 +1,390 @@
+```
+
+2. Resolve any compilation errors by explicitly adding `+ Forget` where needed.
+
+3. Optionally: Recurse into your dependencies, applying the same changes as needed.
+
+Crates using `unsafe` code should beware of `ptr::write` and other unsafe ways of skipping destructors.
+
+## Implications on the libraries
+
+### Relax generic bound on public API
+
+For migrated users it is equivalent to semver's `minor` change, while not migrated uses will observe it as `patch` change.
+
+### Weakening associated type bound and `Self` bound in traits
+
+Bounds for associated types and `Self` in traits would be weakened in respect to the new traits from the start:
+
+```rust
+trait Foo: ?Trait {
+    type Assoc: ?Trait;
+}
+```
+
+This change would not be observable for not migrated crates, because `default_generic_bounds` would default to `Trait`. But if users start migrate before libraries, they will not lock them into old bounds.
+
+```rust
+#![default_generic_bounds(?Forget)]
+
+async fn foo<T: other_crate::Trait>(bar: T) {
+    let fut = bar.baz();
+    // Compiler will emit an error that `fut` maybe `!Forget` because we set `default_generic_bounds`
+    // to `?Forget` and `default_assoc_bounds` in `other_crate` is already `?Forget`. Otherwise it
+    // would have been a breaking change for `other_crate` to make future returned by `baz` `!Forget`,
+    // as this code would've compiled now but not in the future.
+    core::mem::forget(fut);
+}
+
+// Libary that has not migrated yet.
+mod other_crate {
+    trait Trait {
+        async fn baz();
+    }
+}
+```
+
+### Macros
+
+If macro-library generates code, some problems during the migration are possible:
+
+```rust
+mod user {
+    #![default_generic_bounds(?Forget)]
+
+    ::library::make!(); // Will not compile because `T` is `?Forget`.
+}
+
+mod user {
+    #[macro_export]
+    macro_rules! make {
+        () => {
+            pub fn foo<T>(t: T) {
+                ::core::mem::forget(t);
+            }
+        }
+    }
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Introduce new crate level attibute: `default_generic_bounds` used to (non-exhaustively) enumerate overwrides of defaults for different types of bounds. Only a special set of traits would be allowed and would grow with new "breaking" traits, like `Forget`.
+
+Every trait would initally have its unique default. In practice, bounds for all traits that are stable at the date of RFC except `Sized` would default to `?Trait`. For new "breaking" traits, default would be `Trait`, except bounds for `Self` in traits and associated types in traits.
+
+[^trait-not-sized-by-default]: https://rust-lang.github.io/rfcs/0546-Self-not-sized-by-default.html
+
+`default_generic_bounds` is applied for generic parameters. Effectively, it would be observable like that:
+
+```rust
+// crate `b` that has not migrated to `#![default_generic_bounds(?Forget)]`
+mod b {
```

Comment:
This should also mention `dyn Trait`, in argument position I think it's fine  as it essentially acts like a generic bound (so you'd have to imply all the default trait bounds in there but it otherwise works. However, for return position (for example `fn foo() -> Box<dyn Foo>`) it has the same problem as `impl Trait`: it can't be relaxed because existing callers might depend on being able to (for example) forget it.